### PR TITLE
Doc for Python sampling

### PR DESCRIPTION
**Pull request by icantsayrural**
- Added `keanu.algorithm.rst`
- Added docstrings for `sample`, `generate_samples`, `NUTSSampler`, `MetropolisHastingsSampler`
- Removed `distributions.rst` and moved the same content to the top of `keanu.vertex.rst`.
- Made sidebar static.
- New dev dependency `sphinx_autodoc_typehints` so the types are annotated on the parameters not the method signature.

Run `open keanu/docs/python/latest/index.html` locally to view the page.

### CODE REVIEW

File: keanu-python/docs/keanu.vertex.rst [hunk 1]
```diff
@@ -1,6 +1,36 @@
...
+~~~~~~~~~~~~~
+
+* :meth:`keanu.vertex.Bernoulli`
+* :meth:`keanu.vertex.Beta`
+* :meth:`keanu.vertex.Cauchy`
+* :meth:`keanu.vertex.ChiSquared`
+* :meth:`keanu.vertex.Dirichlet`
+* :meth:`keanu.vertex.Exponential`
+* :meth:`keanu.vertex.Gamma`
+* :meth:`keanu.vertex.Gaussian`
```
Comment (wesleyw72) [inline keanu-python/docs/keanu.vertex.rst]: Would you be able to add in `Geometric` distribution to this list?

File: keanu-python/keanu/algorithm/sampling.py [hunk 2]
```diff
@@ -32,6 +32,16 @@ def get_sampler(self) -> JavaObject:
 
 
 class MetropolisHastingsSampler(PosteriorSamplingAlgorithm):
+    """
+    :param proposal_distribution: The proposal distribution for metropolis hastings. Options are 'gaussian' and 'prior'.
```
Comment (wesleyw72) [inline keanu-python/keanu/algorithm/sampling.py]: nit: Can Metropolis Hastings have each first letter capitalised as it's a name of an algorithm.

File: keanu-python/keanu/algorithm/sampling.py [hunk 3]
```diff
@@ -61,6 +71,13 @@ def __init__(self,
 
 
 class NUTSSampler(PosteriorSamplingAlgorithm):
+    """
+    :param adapt_count: The number of samples for which the step size will be tuned. For the remaining samples in which it is not tuned, the step size will be frozen to its last calculated value.
+    :param target_acceptance_prob: The target acceptance probability. A suggested value of this is 0.65 (Beskos et al., 2010; Neal, 2011).
```
Comment (wesleyw72) [inline keanu-python/keanu/algorithm/sampling.py]: Shall we call out what the acceptance probability defaults to if this is left as none?

File: keanu-python/keanu/algorithm/sampling.py [hunk 4]
```diff
@@ -61,6 +71,13 @@ def __init__(self,
 
 
 class NUTSSampler(PosteriorSamplingAlgorithm):
+    """
+    :param adapt_count: The number of samples for which the step size will be tuned. For the remaining samples in which it is not tuned, the step size will be frozen to its last calculated value.
```
Comment (wesleyw72) [inline keanu-python/keanu/algorithm/sampling.py]: Should we say what the adapt count defaults to if this is left as none?

File: keanu-python/keanu/algorithm/sampling.py [hunk 5]
```diff
@@ -61,6 +71,13 @@ def __init__(self,
...
+    :param target_acceptance_prob: The target acceptance probability. A suggested value of this is 0.65 (Beskos et al., 2010; Neal, 2011).
+    :param adapt_enabled: Determines whether the step size will adapt during the first adaptCount samples.
```
Comment (wesleyw72) [inline keanu-python/keanu/algorithm/sampling.py]: What happens when this is None?

File: keanu-python/keanu/algorithm/sampling.py [hunk 6]
```diff
@@ -61,6 +71,13 @@ def __init__(self,
...
+    :param initial_step_size: Sets the initial step size. If none is given then a heuristic will be used to determine a good step size.
+    :param max_tree_height: The maximum tree size for the sampler. This controls how long a sample walk can be before it terminates. This will set at a maximum approximately 2^treeSize number of logProb evaluations for a sample.
```
Comment (wesleyw72) [inline keanu-python/keanu/algorithm/sampling.py]: What happens when this is none?

File: keanu-python/docs/keanu.vertex.rst (hunk 1 above)
Comment (icantsayrural) [reply keanu-python/docs/keanu.vertex.rst]: Yup. This really should be automated somehow

Comment (wesleyw72) [reply keanu-python/docs/keanu.vertex.rst]: I agree, there is a task on the backlog for this.

File: keanu-python/keanu/algorithm/sampling.py [hunk 7]
```diff
@@ -97,6 +114,28 @@ def sample(net: BayesNet,
...
+    :param sampling_algorithm: The posterior sampling algorithm to use.
+        Options are :class:`keanu.algorithm.MetropolisHastingsSampler` and :class:`keanu.algorithm.NUTSSampler`.
+        If not set, :class:`keanu.algorithm.MetropolisHastingsSampler` is chosen with 'prior' as its proposal distribution.
+    :param draws: The number of samples to take.
+    :param drop: The number of samples to drop before collecting anything.
+        If this is zero then no samples will be dropped before collecting.
+    :param down_sample_interval: Collect 1 sample for every `down_sample_interval`.
+        If this is 1 then there will be no down-sampling.
+        If this is 2 then every other sample will be taken.
+        If this is 3 then 2 samples will be dropped before one is taken.
```
Comment (migwellian) [inline keanu-python/keanu/algorithm/sampling.py]: I'd suggest adding "and so on" on a new line, just to make it clear that 1,2,3 are not the only options!

**Pull request merged**